src/Framework/AspNetCoreAnalyzers/src/Analyzers/Mvc/MvcAnalyzer.cs

```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure;
+using Microsoft.AspNetCore.App.Analyzers.Infrastructure;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Analyzers.Mvc;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public partial class MvcAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+        DiagnosticDescriptors.AmbiguousActionRoute
+    );
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(static context =>
+        {
+            var compilation = context.Compilation;
+            var wellKnownTypes = WellKnownTypes.GetOrCreate(compilation);
+            var routeUsageCache = RouteUsageCache.GetOrCreate(compilation);
+
+            // We want ConcurrentHashSet here in case RegisterOperationAction runs in parallel.
+            // Since ConcurrentHashSet doesn't exist, use ConcurrentDictionary and ignore the value.
+            var concurrentQueue = new ConcurrentQueue<List<ActionRoute>>();
+
+            context.RegisterSymbolAction(context =>
+            {
+                if (context.Symbol is INamedTypeSymbol namedTypeSymbol &&
+                    MvcDetector.IsController(namedTypeSymbol, wellKnownTypes))
```

Comment:
The type check is always true and unnecessary.

```suggestion
                var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
                if (MvcDetector.IsController(namedTypeSymbol, wellKnownTypes))
```